docs: added documentation on how to package pipeline for a restricted… (#388)
hydra-genetics/hydra-genetics: docs/packaging_pipeline/prepare_pipeline.md

> # Adjust pipeline

Reviewer: ```suggestion
# Preparing a pipeline for packaging
```

      GRCh37_gencode_v19_CTAT_lib_Mar012021.plug-n-play/ctat_genome_lib_build_dir/ref_annot.prot_info.dbm: 196bf70e87eb0c5640bbebc425319e01
    url: https://figshare.scilifelab.se/ndownloader/files/42084318

fusioncatcher:
  genome_path:
    compressed_checksum: 5006482f5b6b78c99e80330d2d69a239
    path: ref_data/fusioncatcher
    type: folder
    url:
      https://figshare.scilifelab.se/ndownloader/files/42061230: 5f4968cf58a28bfe661ee3d1207259c6
      https://figshare.scilifelab.se/ndownloader/files/42061233: ad677c79617e36f85592b45918a10f21
      https://figshare.scilifelab.se/ndownloader/files/42061242: b4a57d1a48b12e3302fa7d12e6ac6347
      https://figshare.scilifelab.se/ndownloader/files/42061248: 1280e3204b6990b825c762ac14e4446a
    content_checksum:
      human_v102/ucsc_genes_header.txt : 9aa055ae6c632068ddc0f4c60ad9c82f
      human_v102/synonyms.txt : dbaea5a7e374e110b4fbb57c2bea3063
      human_v102/exons_header.txt : 5bdb1d9e7484f7f9ac4666545ad41088
      human_v102/shield_against_pseudo-genes.txt : ae05c5a3a1fc94c8d78442e5f539a47d
```

> ## Modify configs and profile

Reviewer: ```suggestion
## Modifying configs and profile for a new cluster
```

### Resource

> Make a copy of the resources file and make sure it matches your system setup, ex:

Reviewer: ```suggestion
Make a copy of the resources file and make adapt the new resource file to match the new clusters system setup, ex:
```

>  - partition

Reviewer: ```suggestion
 - partition(s)
```

  mem_per_cpu: 6144
  threads: 8

pbrun_fq2bam:
  gres: "--gres=gres:gpu:l40s:2"
  mem_mb: 240000
  mem_per_cpu: 5000
  partition: "gpu"
  threads: 48
```

### Config files

Make new the config files were all file paths point to the reference directory. Example of new config `config/config.data.hg19.miarka.yaml`.
```yaml
# Add the following line at the top with the actual absolute path (example /proj/ngi2024001/bin/wp1_gms560/design_and_ref_files)
REFERENCE_DATA: "<EXTRACT_PATH>/>PIPELINE_SHORT_NAME>/design_and_ref_files"

# Adjust config so that all reference files have the {{REFERENCE_DATA}} variable
# Add potential subfolder so that they match the paths found in the reference yaml files
reference:
  coverage_bed: "{{REFERENCE_DATA}}/hastings/refseq_select_mane_20230828.bed"
  design_bed: "{{REFERENCE_DATA}}/hastings/hg38_exome_comp_spikein_v2.0.2_targets_sorted.re_annotated.sorted.bed"
```

Adjust the main config file to point at the hydra-genetics modules
```yaml
# Add the following line
hydra_local_path: "{EXTRACT_PATH}/{PIPELINE_SHORT_NAME}/{TAG_OR_BRANCH}/hydra-genetics"
```

Add path to local singularities in the main config.yaml file 
```bash
# config/config.yaml

> # Make sure the environment is active

Reviewer: For completeness of the instructions the creation of the python venv followed by the pip install of requirements.txt could be given.

cp config/config.yaml config/config.yaml.copy

> hydra-genetics prepare-environment container-path-update -c config/config.yaml.copy -n config/config.yaml -p ${PATH_TO_singularity_cache}

Reviewer: ```suggestion
hydra-genetics prepare-environment container-path-update -c config/config.yaml.copy -n config/config.yaml -p ${PATH_TO_apptainer_cache}
```

default-resources: [gres="", ]
drmaa-log-dir: "slurm"
singularity-args: "-e --cleanenv -B /proj -B $HOME --nv"
singularity-prefix: "/proj/ngi2024001/bin/<PIPELINE_SHORT_NAME>/<TAG_OR_BRANCH>/singularity_cache/"
wrapper-prefix: "git+file://proj/ngi2024001/bin/<PIPELINE_SHORT_NAME>/<TAG_OR_BRANCH>/snakemake-wrappers/"
```

## Modify Snakemake module import
To be able to use a local git file module imports must use the `get_module_snakefile()` function (instead of `github()`).

```py
module alignment:
    snakefile:
        get_module_snakefile( # Modify this line for all module imports
            config,
            "hydra-genetics/alignment",
            path="workflow/Snakefile",
            tag=config["modules"]["alignment"],
        )
    config:
        config
```

## Modify common.smk
Add these line in `rules/common.smk`:

```py
from hydra_genetics.utils.misc import get_module_snakefile # Needed for local git
from hydra_genetics.utils.misc import replace_dict_variables # Used by hydra genetics to obtain the file paths to the files in the config

config = replace_dict_variables(config) # Add this after the imports
```

## Add build script to pipeline

> Add the build script `build/build_conda.sh` found in `https://github.com/hydra-genetics/hydra-genetics` to the pipeline.

Reviewer: ```suggestion
Add the build script `build/build_conda.sh` found in `https://github.com/hydra-genetics/hydra-genetics` to the pipeline. This can be placed in a subdirectory of your pipeline call build
```

```

Add path to local singularities in the main config.yaml file 
```bash
# config/config.yaml
# Build new python environment if needed and install requirements
python3.11 -m venv python_venv/
pip install -r requirements.txt
# Activate python environment if not already active
source python_venv/bin/activate
cp config/config.yaml config/config.yaml.copy
hydra-genetics prepare-environment container-path-update -c config/config.yaml.copy -n config/config.yaml -p ${PATH_TO_apptainer_cache}
```

### Profile

Make a new profile and adjust it to the computer cluster 
```bash
cp profile/marvin/config.yaml profile/miarka/config.yaml
```

```yaml
# profiles/miarka/config.yaml
# Example
jobs: 100
keep-going: True
restart-times: 1
rerun-incomplete: True
max-jobs-per-second: 100
max-status-checks-per-second: 100
use-singularity: True
drmaa: "-p {resources.partition} -t {resources.time} -n {resources.threads} --mem={resources.mem_mb} --mem-per-cpu={resources.mem_per_cpu} {resources.gres} -J {rule} -A ngi2024001 -e slurm/{rule}_%j.err -o slurm/{rule}_%j.out --nodes=1-1"
default-resources: [gres="", ]
drmaa-log-dir: "slurm"
singularity-args: "-e --cleanenv -B /proj -B $HOME --nv"

> singularity-prefix: "/proj/ngi2024001/bin/<PIPELINE_SHORT_NAME>/<TAG_OR_BRANCH>/singularity_cache/"

Author: ```suggestion
singularity-prefix: "/proj/ngi2024001/bin/<PIPELINE_SHORT_NAME>/<TAG_OR_BRANCH>/apptainer_cache/"
```

> singularity-prefix: "/proj/ngi2024001/bin/<PIPELINE_SHORT_NAME>/<TAG_OR_BRANCH>/singularity_cache/"

Author: ```suggestion
singularity-prefix: "/proj/ngi2024001/nobackup/bin/<PIPELINE_SHORT_NAME>/<TAG_OR_BRANCH>/apptainer_cache/"
```
